### DISCUSSION TITLE
App1(host) include App2(remote), router is not working

### DISCUSSION THREAD

**Discussion by franklion** [originjs/vite-plugin-federation#195] [Q&A] [answered]
App1:host, vue3, simple application, just show h1 APP1
App2:remote, vue3, **with router and store**

App1 include App2, use module-federation.

when i execute App1, get a warn, and router is not working :( 

> [Vue warn]: Failed to resolve component: router-link
> If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement.

> [Vue warn]: Failed to resolve component: router-view
> If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement. 

![截圖 2022-06-17 下午11 43 19](https://user-images.githubusercontent.com/10952687/174333583-133ce8f5-a143-4eeb-bc25-4de11d675a7a.png)

So if App2 is remote module, can it use router?

- app1 App.vue
```javascript
<script>
import { defineAsyncComponent } from "vue";

export default {
  name: "App",
  components: {
    RemoteApp: defineAsyncComponent(() => import("app2/App")),
  },
};
</script>

<template>
  <h1>APP1</h1>
  <RemoteApp></RemoteApp>
</template>
```

- app1 vite.config.js
```javascript
export default defineConfig({
  plugins: [
    vue(),
    federation({
      name: "app1",
      filename: "remoteEntry.js",
      remotes: {
        app2: "http://localhost:4173/assets/remoteEntry.js",
      },
      shared: ["vue", "vue-router", "vuex"],
    }),
  ],
  resolve: {
    alias: {
      "@": fileURLToPath(new URL("./src", import.meta.url)),
    },
  },
});
```

- app2 vite.config.js
```javascript
export default defineConfig({
  plugins: [
    vue(),
    federation({
      name: "app2",
      filename: "remoteEntry.js",
      exposes: {
        "./App": "./src/App.vue",
      },
      shared: ["vue", "vue-router", "vuex"],
    }),
  ],
  resolve: {
    alias: {
      "@": fileURLToPath(new URL("./src", import.meta.url)),
    },
  },
  build: {
    target: "esnext",
    minify: false,
    cssCodeSplit: true,
    rollupOptions: {
      output: {
        minifyInternalExports: false,
      },
    },
  },
});
```

github-link: [https://github.com/franklion/vue-module-federation](https://github.com/franklion/vue-module-federation)

**Comment by Jiannan-dev** [comment, marked as answer]
Thanks for your feedback, I guess the reason for this problem is the same as the previous issue (https://github.com/originjs/vite-plugin-federation/issues/139#issuecomment-1060254876), the router should also be bound to the vue instance, remote and host are running in different vue instances, so they can't share the router, when you need to define the router, you have to define it on the host side, like this
``` js
const router = createRouter({
    history: createWebHashHistory(), 
    routes: [
        {
            path: '/',
            redirect: '/introduce'
        },
        {
            path: '/login',
            name: 'login',
            component: defineAsyncComponent(() => import("router-remote/Login"))
        },
        {
            path: '/shopping',
            name: 'shopping',
            component: () => import( '../views/Shopping.vue')
        }
        ,
        {
            path: '/account',
            name: 'account',
            component: () => import('../views/Account.vue')
        }
    ]
})
```

**Reply by franklion** [reply]
ok, I see. Thank you :)

By the way, if the host module is Angular or React, can i use remote module (Vue) with router?
If I want to use router, does it have to be used on the host side?

**Reply by Jiannan-dev** [reply]
> By the way, if the host module is Angular or React, can i use remote module (Vue) with router?

Do you mean using Vue components in React or Angular? This is unlikely, plugins are not compatible with the differences between frameworks, if you only use Vue on the remote side (it does not need to be used in the host), then you can use the components as you wish

**Reply by Jiannan-dev** [reply]
> If I want to use router, does it have to be used on the host side?

The use of router is not limited by host and remote, because it is bound to different vue instances, that is, if the component needs to run in host, then you need to define router in host, if you need to run in remote, then you need to define router in remote, these two do not conflict, you can set in remote and host for the same component in different router

**Reply by ghost** [reply]
This allows to have host routing and internal routing? Thanks.
